### ISSUE TITLE
How to handle Surfaces part of a component but not an instance?

### ISSUE THREAD

**Issue by Kevinator123** [moethu/SketchUpNET#18]
Each vertice that is part of a surface which is part of an instance you can use .GetTransformed on and you get the correct position of the vertice to build your mesh but some components aren't part of an instance and contain surfaces that aren't part of an instance so how do you go about positioning those vertices in the correct place? 

Sorry I have to ask it here but I don't know where else to go, the documentation on SketchUp file structures is very limited.

**Comment by moethu** [comment]
Hi @Kevinator123 so you want to get all surfaces as one list, correct?
So my take would be:
1. get all surfaces into a list
2. walk through all instances and get the surfaces and add them to the list
like this one:
https://github.com/moethu/SketchUpNET/blob/master/SketchUp/SketchUpNET/SketchUpForDynamo/SketchUp.cs#L280-L308

**Comment by Kevinator123** [comment]
Thanks for answering, the way I do it right now is check each instance, surface and group on model level and keep on going down the nested components but I'm still missing faces somehow. 

https://github.com/Kevinator123/SketchUpReader/blob/master/ReadSKPConsole/Program.cs

**Comment by Kevinator123** [comment]
I think I checked all nested surfaces, foreach instance I get the parent component, for each of those components I draw the surfaces, I check the groups and check the instances. For each group I check the instances and their parent components and nested groups and for components I draw the surfaces, check for new instances and new groups. I also check for surfaces, groups and components on model level.

**Comment by Kevinator123** [comment]
It definitely are just faces so I think it can be meshed, they are part of an instance. But what I just noticed is that everything that doesn't draw has as type: Type undefined. Perhaps that has something to do with it?

**Comment by moethu** [comment]
Type undefined sounds odd, can you share your skp model with me so I can have a look. And please point me to the failing faces.

**Comment by 13704017891** [comment]
@ KevinHuyskens Hello, do you import SKP with unity?I am doing the same thing, can you give me some reference?

**Comment by moethu** [comment]
@KevinHuyskens , sorry just realizing you answered quite some time ago. I was wondering when I looked at your code: what about instances that contain other instances? and what about components containing components? both cases are not catched.

**Comment by KevinBME** [comment]
@13704017891 Yes, I load the file externally and pass all the faces through with a named pipe. (haven't worked on this project for quiete some time).
@moethu I'll try this when I have time, see if it works, I'll keep you posted once I get to it. 

(Same person, different account)

**Comment by KevinHuyskens** [comment]
@moethu took a look at your suggestion just now, but your library's Instance  object doesn't contain any Instances and Component doesn't contain any Components so I can't catch those cases :')

**Comment by moethu** [comment]
Did you check the Dynamo implementation as a reference? The components in there were able to catch all geometries so far. https://github.com/moethu/SketchUpNET/blob/master/SketchUp/SketchUpNET/SketchUpForDynamo/SketchUp.cs
Maybe I was missing something, I will setup a Windows VM to test you implementation but please have a look if you got all the nested elements etc.

**Comment by KevinHuyskens** [comment]
So instead of going over my old code I started completely over, here is a code sample for writing out everything in a treeview in wpf (I excluded edges and curves because I don't need them). 

Here is the sample: https://pastebin.com/XpRWYruK
(start a new wpf project and add a treeview with name "treeView" if you want to use it yourself and change the model path and perhaps the namespace.)

Redoing it I can already see I missed some essential parts in my previous approach so I'll let you know how it goes from here on, but I've a feeling that I'm on the right track ;)

**Comment by moethu** [comment]
Let me know if anything is missing in the library
